### PR TITLE
feat(airbyte-cdk): CSVExtractor

### PR DESCRIPTION
**Pull request by natikgadzhi**
## What

A CSV record extractor for APIs that return a CSV string in response text! [Someone on Community Slack asked and I was curious](https://airbytehq-team.slack.com/archives/C027KKE4BCZ/p1720161452667559).

## How

- Added a CSV extractor as a valid record extractor type in the declarative component schema
- Added the model to type mapping
- Added the `CSVExtractor` implementation

## Review guide
This is PoC grade, still missing tests

## User Impact

Once merged and released, uses will be able to specify CSVExtractor instead of DPathExtractor in their manifests, but it won't yet be available in the Builder UI.

> [!warning]
> This is a minor CDK Version bump

## Can this PR be safely reverted and rolled back?
- [x] YES 💚
- [ ] NO ❌

## TODO
- [ ] Add unit tests
- [ ] Make a local example using local CDK in Builder via custom yaml component, verify it works
- [ ] Consider adding the UI components in a separate PR

## Summary by CodeRabbit

- **New Features**
  - Introduced `CSVExtractor` for enhanced CSV data extraction capabilities with customizable separator and quote character options.
  
- **Chores**
  - Updated `.gitignore` to exclude the `dist` directory of the Airbyte CDK Python component.

### CODE REVIEW

**Comment by natikgadzhi** [comment]
People, error output of airbyte-ci checks here does not look helpful: 

```
21: exec pip install datamodel_code_generator==0.11.19 DONE

20: mkdir /generated
20: mkdir /generated DONE

27: mkfile /generated/__init__.py DONE
27: mkfile /generated/__init__.py DONE

26: exec datamodel-codegen --input /yaml/declarative_component_schema.yaml --output /generated/declarative_component_schema.py --disable-timestamp --enum-field-as-literal one --set-default-enum-member DONE
26: exec datamodel-codegen --input /yaml/declarative_component_schema.yaml --output /generated/declarative_component_schema.py --disable-timestamp --enum-field-as-literal one --set-default-enum-member DONE

25: exec mkdir /generated_post_processed
25: exec mkdir /generated_post_processed DONE

12: resolve image config for docker.io/library/python:3.10
12: > in from python:3.10
12: resolve image config for docker.io/library/python:3.10 DONE

12: resolve image config for docker.io/library/python:3.10 DONE
12: > in from python:3.10
12: resolve image config for docker.io/library/python:3.10 DONE

34: mkdir /generated_post_processed
34: mkdir /generated_post_processed DONE

33: mkfile /generated_post_processed/__init__.py
33: mkfile /generated_post_processed/__init__.py DONE

12: resolve image config for docker.io/library/python:3.10
12: > in from python:3.10
12: resolve image config for docker.io/library/python:3.10 DONE

39: mkdir /generated_post_processed
39: mkdir /generated_post_processed DONE

38: mkfile /generated_post_processed/declarative_component_schema.py
38: mkfile /generated_post_processed/declarative_component_schema.py DONE

41: export directory /generated_post_processed to host airbyte_cdk/sources/declarative/models
41: ...

42: copy /generated_post_processed / DONE
42: copy /generated_post_processed / DONE

41: export directory /generated_post_processed to host airbyte_cdk/sources/declarative/models DONE
41: export directory /generated_post_processed to host airbyte_cdk/sources/declarative/models DONE
NoneType: None
airbyte-cdk/python: Attempting to create failure status for commit [641](https://github.com/airbytehq/airbyte/actions/runs/9822869183/job/27120063649?pr=41016#step:6:656)51965e0f5e8ac45142efffc6ddb5bda75b6c0 on Github in airbyte-cdk/python - check-ci context.
airbyte-cdk/python: Created failure status for commit 64151965e0f5e8ac45142efffc6ddb5bda75b6c0 on Github in airbyte-cdk/python - check-ci context with desc: check-ci execution for airbyte-cdk/python.
Error: The following packages failed to run poe tasks:  airbyte-cdk/python
Sentry is attempting to send 1 pending events
Waiting up to 2 seconds
Press Ctrl-C to quit
Error: Process completed with exit code 1.
Run mapfile -t containers < <(docker ps --filter name="dagger-engine-*" -q)
5765ee109ea3
Run jwalton/gh-docker-logs@v2
Found 1 containers...
Found 1 matching containers...
```

How do you work with this? I _think_ this is mypy and will try and fix it, but.

**Comment by natikgadzhi** [comment]
@coderabbit review

**Comment by natikgadzhi** [comment]
@coderabbitai review

**Comment by natikgadzhi** [comment]
@artem1205 I think I might have done this wrong — how does it look to you?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/extractors/csv_extractor.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        separator (str): Record separator in the incoming CSV.
+        quotechar (str): Quotation character used in the CSV.
+    """
+
+    separator: str
+    quotechar: str
+
+    def extract_records(self, response: requests.Response) -> Iterable[Mapping[str, Any]]:
+        try:
+            csv_data = StringIO(response.text)
```
Comment (artem1205) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/extractors/csv_extractor.py:42]: Despite the fact you've created the buffer here, [response body is still downloaded in full](https://requests.readthedocs.io/en/latest/user/advanced/#body-content-workflow) in httpclient (it is default behaviour, unless `stream=True` s is passed).

**Review by artem1205 — CHANGES_REQUESTED** [review]
Looks good, but another component should be used here.

So, what should be done?

new `Decoder` instead of `RecordExtractor`:
1. `class CSVDecoder(Decoder)` should be introduced  with stream option enabled.
2. pandas.read_csv can be used to read , see [salesforce](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py#L543-L547)

**Comment by natikgadzhi** [comment]
@artem1205 thank you for the review! <3 

I'm not _actively_ looking here, but I'll try and wrap this up when I have an hour.

**Comment by natikgadzhi** [comment]
UPD: not actively in progress, but I'll clean up when I have time.

**Comment by natikgadzhi** [comment]
Update here: I'm pretty determined to get this to work, esp as @pnilan works on a related things in XML. I'll try to sit down with Patrick to pair this week, but if not, we'll do this during the cycle.

**Comment by natikgadzhi** [comment]
Closing in favor of #47131

**Pull request closed without merging**